### PR TITLE
Trivial optimize of function coalesce. 

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reuse the result of `FunctionFactory::instance().get("isNotNull", context)` and `FunctionFactory::instance().get("assumeNotNull", context)`. Make sure it is called once during the lifetime of `FunctionCoalesce`. 

Notice that the cost of `FunctionFactory::instance().get` can't be ignored during the execution of function `coalesce`. In my tests: 
- `executeImpl` costs 68068 ns
- `FunctionFactory::instance().get` inside `executeImpl` costs 9537 ns, which is 14% of `executeImpl`

``` sql 
select coalesce(materialize(null), -1) from numbers(1000000000) format Null settings max_block_size = 8192;  
Old: 0 rows in set. Elapsed: 2.478 sec. Processed 1.00 billion rows, 8.00 GB (403.61 million rows/s., 3.23 GB/s.)
New: 0 rows in set. Elapsed: 2.321 sec. Processed 1.00 billion rows, 8.00 GB (430.86 million rows/s., 3.45 GB/s.)
```

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 30a355d9d49c103f81412ddbf8fa2372aa15b3a6 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59627/30a355d9d49c103f81412ddbf8fa2372aa15b3a6/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59627/30a355d9d49c103f81412ddbf8fa2372aa15b3a6/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Functions/coalesce.cpp [hunk 1]
```diff
@@ -110,8 +115,7 @@ class FunctionCoalesce : public IFunction
                 break;
         }
 
-        auto is_not_null = FunctionFactory::instance().get("isNotNull", context);
-        auto assume_not_null = FunctionFactory::instance().get("assumeNotNull", context);
+
```
Comment (antaljanosbenjamin) [inline src/Functions/coalesce.cpp:118]:
```suggestion
```

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

**Review by CurtizJ — COMMENTED** [review]
You can also optimize https://github.com/ClickHouse/ClickHouse/blob/5a789f5387fd5d81166431e46be81970021101bd/src/Functions/coalesce.cpp#L154-L156

**Review by antaljanosbenjamin — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Comment by antaljanosbenjamin** [comment]
The test failures are not related, I will merge the PR.

**Pull request merged**